### PR TITLE
2*c connection/goroutine leak fix (std client/ui)

### LINKED ISSUE #166: 2*c connection leaks when using the UI to trigger run (regression?)
reloading many times while testing histogram boundary:

http://localhost:8080/fortio/?labels=Fortio&url=http%3A%2F%2Flocalhost%3A8080%2Fecho%3Fdelay%3D1ms%3A90%2C9ms%3A10&qps=100&t=&t=on&n=100&c=10&p=50%2C+75%2C+90%2C+99%2C+99.9&r=0.0000001&H=User-Agent%3A+istio%2Ffortio-0.7.2-pre&H=&H=&runner=http&save=on&load=Start

Starting at 100 qps with 10 thread(s) [gomax 4] : exactly 100, 10 calls each (total 100 + 0)
URL: localhost:8080/echo?delay=1ms:90,9ms:10
```
23:32:05 I uihandler.go:439> UI: GET /fortio/?labels=Fortio&url=http%3A%2F%2Flocalhost%3A8080%2Fecho%3Fdelay%3D1ms%3A90%2C9ms%3A10&qps=100&t=&t=on&n=100&c=10&p=50%2C+75%2C+90%2C+99%2C+99.9&r=0.0000001&H=User-Agent%3A+istio%2Ffortio-0.7.2-pre&H=&H=&runner=http&save=on&load=Start HTTP/1.1 [::1]:61957 ()
23:32:05 I uihandler.go:153> Starting http load request from [::1]:61957 for http://localhost:8080/echo?delay=1ms:90,9ms:10
23:32:05 I uihandler.go:209> New run id 13
23:32:05 I uihandler.go:227> Estimating fixed #call 100 duration to 1 seconds 100
23:32:05 I httprunner.go:72> Starting http test for http://localhost:8080/echo?delay=1ms:90,9ms:10 with 10 threads at 100.0 qps
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 http: Accept error: accept tcp [::]:8080: accept: too many open files; retrying in 5ms
23:32:05 E http.go:663> Read error &{{0xc420e35b00}} {::1 8080 } 0 : read tcp [::1]:62091->[::1]:8080: read: connection reset by peer
23:32:05 http: Accept error: accept tcp [::]:8080: accept: too many open files; retrying in 5ms
23:32:05 E http.go:663> Read error &{{0xc420e35b80}} {::1 8080 } 0 : read tcp [::1]:62092->[::1]:8080: read: connection reset by peer
23:32:05 http: Accept error: accept tcp [::]:8080: accept: too many open files; retrying in 5ms
23:32:05 E http.go:663> Read error &{{0xc4206a1980}} {::1 8080 } 0 : read tcp [::1]:62094->[::1]:8080: read: connection reset by peer
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
23:32:05 E http.go:556> Unable to connect to {::1 8080 } : dial tcp [::1]:8080: socket: too many open files
```

### PR DESCRIPTION
**Pull request by ldemailly**
- [x] Phase 1 : repro with tests failing
- [x] Phase 2: code fix

Also verified manually with lsof

Fixes #166

### CODE REVIEW

File: Webtest.sh [hunk 1]
```diff
@@ -23,6 +23,8 @@ $CURL https://istio.io/robots.txt
 $CURL ${BASE_FORTIO}browse
 # Check we can connect, and run a http QPS test against ourselves through fetch
 $CURL "${BASE_FORTIO}fetch/localhost:8080$FORTIO_UI_PREFIX?url=http://localhost:8080/debug&load=Start&qps=-1&json=on" | grep ActualQPS
+# Check we can do it twice despite ulimit - check we get all 200s (exactly 80 of them (default is 8 connections->16 fds + a few))
+$CURL "${BASE_FORTIO}fetch/localhost:8080$FORTIO_UI_PREFIX?url=http://localhost:8080/debug&load=Start&n=80&qps=-1&json=on" | grep '"200": 80'
```
Comment (asadali) [inline Webtest.sh:27]: +1

File: fhttp/httprunner_test.go [hunk 2]
```diff
@@ -72,6 +73,57 @@ func TestHTTPRunner(t *testing.T) {
 	}
 }
 
+func testHTTPNotLeaking(t *testing.T, opts *HTTPRunnerOptions) {
+	ngBefore1 := runtime.NumGoroutine()
+	t.Logf("Number go rountine before test %d", ngBefore1)
```
Comment (asadali) [inline fhttp/httprunner_test.go]: s/rountine/routine/g

File: fhttp/httprunner_test.go [hunk 3]
```diff
@@ -72,6 +73,57 @@ func TestHTTPRunner(t *testing.T) {
...
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	httpOk = res.RetCodes[http.StatusOK]
+	if opts.Exactly != httpOk {
+		t.Errorf("Run2: Mismatch between requested calls %d and ok %v", numCalls, res.RetCodes)
+	}
+	// allow for ~5 goroutine variance, as we use 50 if we leak it will show
+	if ngAfter > ngBefore2+5 {
```
Comment (asadali) [inline fhttp/httprunner_test.go]: +1 TIL how to test goroutine leakages

**Review by kyessenov — COMMENTED** [review]
Looks reasonable.
Have you tried running it at 30-60kqps and observing system socket states?

**Review by asadali — APPROVED** [review]
thanks for detecting and solving this! I think I may have been running into it over long deploys of fortio.

detected a minor typo but looks good to go in

File: fhttp/httprunner_test.go (hunk 2 above)
Comment (asadali) [reply fhttp/httprunner_test.go]: maybe also change to `Number of`

**Comment by ldemailly** [comment]
@kyessenov 

> Have you tried running it at 30-60kqps and observing system socket states?

my mac can only do 44k qps with 48 connections

```
$ lsof -p 98312|grep TCP|wc -l
     101
```
stable

```
All done 2658061 calls 1.098 ms avg, 43709.8 qps
```

**Pull request merged**